pkg/collector/corechecks/snmp/internal/profile/rc_provider.go

```diff
@@ -0,0 +1,118 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+package profile
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/DataDog/datadog-agent/comp/remote-config/rcclient"
+	"github.com/DataDog/datadog-agent/pkg/networkdevice/profile/profiledefinition"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"maps"
+	"slices"
+	"sync"
+	"time"
+)
+
+var rcSingleton *UpdatableProvider
+var rcOnce sync.Once
+var rcError error
+
+// ResetRCProvider destroys the singleton instance. This should only be used in tests.
+// TODO Turn this back into a proper Component so that we can mock it instead of using a singleton.
+func ResetRCProvider() {
+	rcSingleton = nil
+	rcOnce = sync.Once{}
+	rcError = nil
+}
+
+// NewRCProvider returns a profile provider that subscribes to remote
+// configuration and receives profile updates from the backend. Multiple calls
+// will return the same singleton object.
+func NewRCProvider(client rcclient.Component) (Provider, error) {
+	rcOnce.Do(func() {
+		rcSingleton, rcError = buildAndSubscribeRCProvider(client)
+	})
+	return rcSingleton, rcError
+}
+
+// buildAndSubscribeRCProvider builds a new UpdatableProvider and subscribes to
+// RC to receive profile updates.
+func buildAndSubscribeRCProvider(rcClient rcclient.Component) (*UpdatableProvider, error) {
+	// Load OOTB profiles from YAML
+	defaultProfiles := getYamlDefaultProfiles()
+	if defaultProfiles == nil {
+		return nil, fmt.Errorf("could not find OOTB profiles")
+	}
+	userProfiles := make(ProfileConfigMap)
+
+	provider := &UpdatableProvider{}
+	provider.Update(userProfiles, defaultProfiles, time.Now())
+
+	// Subscribe to the RC client
+	log.Info("Subscribing to remote config for device profiles")
+	rcClient.Subscribe(state.ProductNDMDeviceProfilesCustom, makeOnUpdate(provider))
+
+	return provider, nil
+}
+
+// unpackRawConfigs converts a map of raw remote config data to a map of parsed
+// profiles.
+func unpackRawConfigs(update map[string]state.RawConfig) (ProfileConfigMap, map[string]error) {
+	errors := make(map[string]error)
+	profiles := make(ProfileConfigMap)
+	// iterate over keys in sorted order for determinism
+	keys := slices.Sorted(maps.Keys(update))
+	for _, k := range keys {
+		v := update[k]
+		var def profiledefinition.DeviceProfileRcConfig
+		err := json.Unmarshal(v.Config, &def)
+		if err != nil {
```

Comment:
That's a nit, but here it's a good candidate to the idiomatic (especially since you use that form L80 for the map check):
```suggestion
		if err := json.Unmarshal(v.Config, &def); err != nil
```

Not a strong opinion.